### ISSUE TITLE
Null pointers for cached intensive quantities in overlap region.

### ISSUE THREAD

**Issue by blattms** [OPM/ewoms#158]
Just did a test run of SPE1 in parallel and I get segmentation faults because in [SimulatorFullyImplicitBlackoilEbos.hpp#L668](/OPM/opm-simulators/blob/81ebba2df802ea55b40f1892619f1d00ab6f2920/opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp#L668) the intensive quantities are null pointers for some cells (presumable non-interior cells).

Has this changed/ is expected?

**Comment by andlaus** [comment]
the linearization procedure does not seem to need these quantities, so relying on their presence after linearization is bound to fail. As I see it, the opm-simulators code should be uniformly converted to the generic loop over grids to avoid this. (like already done e.g. here: [BlackoilModelEbos.hpp#L1015](https://github.com/OPM/opm-simulators/blob/81ebba2df802ea55b40f1892619f1d00ab6f2920/opm/autodiff/BlackoilModelEbos.hpp#L1015))

**Comment by blattms** [comment]
I could not disagree more.

1. Quantities/Containers behaving differently on parts of grid will be a surprise to most developers. This is even more an issue as this seems to be documented nowhere.
2. Having to adjust algorithms depending on which partitions an element is one clutters the code and demands even more knowledge of of what a parallel DUNE grid is. This is a higher hurdle for new developers.
3. One needs a lot more code with your proposal.
```
ElementContext elemCtx(ebosSimulator_);
const auto& elemEndIt = elemCtx.gridView().template end</*codim=*/0>();
for (auto elemIt = elemCtx.gridView().template begin</*codim=*/0>();
       elemIt != elemEndIt;
      ++elemIt)
{
       const auto& elem = *elemIt;
       if (elem.partitionType() != Dune::InteriorEntity) {
           continue;
       }

elemCtx.updatePrimaryStencil(elem);
elemCtx.updatePrimaryIntensiveQuantities(/*timeIdx=*/0);
const auto& intQuants = elemCtx.intensiveQuantities(/*spaceIdx=*/0, /*timeIdx=*/0);
```
instead of
```
for (int cellIdx = 0; cellIdx < numCells; ++cellIdx) {
    const auto& intQuants = *ebosSimulator_.model().cachedIntensiveQuantities(cellIdx, /*timeIdx=*/0);
```

BTW these surprises cost a lot of time.

**Comment by andlaus** [comment]
> I could not disagree more.

that's fine with me, just fix the fallout behind you then ;)

**Comment by andlaus** [comment]
regarding the amount of code: that is significantly reduced if `Dune::CpGrid` supported looping over given partitions and if `elemCtx.updateAll(elem)` is used. the latter just happens to be slower because it updates everything which is needed to linearize the element. This would look like this:

```c++
for (const auto&& elem: gridView.template elements<Dune::InteriorPartition>()) {
  elemCtx.updateAll(elem);
  const auto& intQuants = elemCtx.intensiveQuantities(0, 0);
}
```

(not really sure what the exact Dune interface is.)

> BTW these surprises cost a lot of time.

in this instance, nothing surprising would have happened if generic way to loop over grids was used  uniformly...

**Comment by andlaus** [comment]
since this is not an eWoms issue and it is fixed by OPM/opm-simulators#1095, I'm closing this.
